nengo_fpga/dna_extractor.py

```diff
@@ -0,0 +1,255 @@
+                    got_error = 2
+
+                if info_str.startswith('Traceback'):
+                    print('<%s> ENCOUNTERED ERROR!' % remote_ip, flush=True)
+                    got_error = 1
+                elif got_error > 0 and info_str[0] != ' ':
+                    # Error string is no longer tabbed, so the actual error
+                    # is bring printed. Collect and terminate (see below)
+                    got_error = 2
+
+                if got_error > 0:
+                    # Once an error is encountered, keep collecting error
+                    # messages until the termination condition (above)
+                    error_strs.append(info_str)
+                else:
+                    print('<%s> %s' % (remote_ip, info_str), flush=True)
+            self.ssh_info_str = info_str_list[-1]
+
+            # The traceback usually contains 3 lines, so collect the first
+            # three lines then display it.
+            if got_error == 2:
+                ssh_channel.close()
+                raise RuntimeError(
+                    'Received the following error on the remote side <%s>:\n%s'
+                    % (remote_ip, '\n'.join(error_strs)))
+
+    def connect(self):
+        print("<%s> Open SSH connection" %
+              fpga_config.get(self.fpga_name, 'ip'), flush=True)
+        # Start a new thread to open the ssh connection. Use a thread to
+        # handle the opening of the connection because it can lag for certain
+        # devices, and we dont want it to impact the rest of the build process.
+        connect_thread = threading.Thread(target=self.connect_thread_func,
+                                          args=())
+        connect_thread.start()
+
+    def process_ssh_output(self, data):
+        # Clean up the data stream coming back over ssh
+        str_data = data.decode('latin1').replace('\r\n', '\r')
+        str_data = str_data.replace('\r\r', '\r')
+        str_data = str_data.replace('\r', '\n')
+
+        # Process and dump the returned ssh data to logger. Data (strings)
+        # returned over SSH are terminated by a newline, so, keep track of
+        # the data and write the data to logger only when a newline is
+        # received.
+        self.ssh_info_str += str_data
+
+    @property
+    def ssh_string(self):
+        # Generate the string to be sent over the ssh connection to run the
+        # remote side ssh script (with appropriate arguments)
+        if self.config_found:
+            ssh_str = \
+                ('python ' + fpga_config.get(self.fpga_name, 'dna_script') +
+                 ' --host_ip="%s"' % fpga_config.get('host', 'ip') +
+                 ' --tcp_port=%i' % self.tcp_port +
+                 '\n')
+        else:
+            ssh_str = ''
+        return ssh_str
+
+    def recv_DNA(self):
+        # Try to connect to FPGA socket a few times
+        for i in range(self.loss_limit):
```

Comment:
Thanks for the explanation! this makes sense.

I also changed the variable names from `loss_limit` and `packet_losses` to `max_attempts` and `connect_attempts` since this is TCP, technically we are just attempting to connect, we aren't losing packets.

Reply:
👍